File: pubs.pl
#!/usr/bin/perl -w

use strict;


###########################################
use Getopt::Long;

my ($HELP, $VERBOSE) = (   "",     0);
my ($MODE) = ("time");

GetOptions ("help"         => \$HELP,
            "verbose!"     => \$VERBOSE,
	    "mode=s"       => \$MODE,
           );
if ($HELP or @ARGV == 0) {usage(); exit(0);}


sub usage {
  print STDERR "\n usage: $0 [options] <list of files>
  --help: this message
  --verbose: debug mode [off]
  --mode <mode>: one of {time, topic, selected, cv} 
\n";
}

my @files = @ARGV;
##################################

my @topicOrder;
my %topics; 

my $count=-1;

my @allEntries;


#############################


sub PrintHeader() {

    print '
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Ratul Mahajan - Publications</title>
  <style>
    * {
      margin: 0;
      padding: 0;
      box-sizing: border-box;
    }
    
    body {
      font-family: -apple-system, BlinkMacSystemFont, \'Segoe UI\', Roboto, Oxygen, Ubuntu, Cantarell, sans-serif;
      line-height: 1.6;
      color: #333;
      background: #fff;
      padding: 20px;
    }
    
    .container {
      max-width: 900px;
      margin: 0 auto;
      padding: 40px 20px;
    }
    
    h1 {
      font-size: 2.5em;
      font-weight: 300;
      margin-bottom: 30px;
      color: #1a1a1a;
      border-bottom: 2px solid #0066cc;
      padding-bottom: 15px;
    }
    
    h2 {
      font-size: 1.8em;
      font-weight: 400;
      margin: 40px 0 20px 0;
      color: #1a1a1a;
    }
    
    .nav-links {
      margin-bottom: 40px;
      padding: 20px;
      background: #f8f9fa;
      border-radius: 8px;
    }
    
    .nav-links ul {
      list-style: none;
      display: flex;
      flex-wrap: wrap;
      gap: 20px;
    }
    
    .nav-links li:before {
      content: "→";
      margin-right: 8px;
      color: #0066cc;
    }
    
    .nav-links a {
      color: #0066cc;
      text-decoration: none;
      font-weight: 500;
    }
    
    .nav-links a:hover {
      text-decoration: underline;
    }
    
    table {
      width: 100%;
      border-collapse: collapse;
      margin-bottom: 30px;
    }
    
    td {
      padding: 15px 0;
      vertical-align: top;
    }
    
    td:first-child {
      width: 10px;
    }
    
    a {
      color: #0066cc;
      text-decoration: none;
      transition: color 0.2s;
    }
    
    a:hover {
      color: #004499;
      text-decoration: underline;
    }
    
    b {
      font-weight: 600;
      color: #1a1a1a;
    }
    
    .year-header {
      font-size: 1.5em;
      font-weight: 500;
      color: #0066cc;
      margin-top: 40px;
      margin-bottom: 20px;
      padding-bottom: 10px;
      border-bottom: 1px solid #e0e0e0;
    }
    
    .paper-entry {
      margin-bottom: 25px;
      padding-bottom: 25px;
      border-bottom: 1px solid #f0f0f0;
    }
    
    .paper-entry:last-child {
      border-bottom: none;
    }
    
    .paper-title {
      font-size: 1.05em;
      font-weight: 600;
      margin-bottom: 6px;
      line-height: 1.4;
    }
    
    .paper-authors {
      color: #666;
      margin: 4px 0;
    }
    
    .paper-venue {
      color: #888;
      font-style: italic;
      margin: 4px 0;
    }
    
    .paper-note {
      color: #0066cc;
      font-weight: 600;
      margin-top: 6px;
    }
    
    .paper-resource {
      margin-top: 6px;
    }
    
    @media (max-width: 768px) {
      h1 {
        font-size: 2em;
      }
      
      .nav-links ul {
        flex-direction: column;
        gap: 10px;
      }
    }
  </style>
</head>
<body>
  <div class="container">
    <h1>Publications</h1>
';
}


sub PrintFooter() {

    print "
  </div>
</body>
</html>
";

}

sub PrintByTopic()  {

    print "<div class=\"nav-links\">\n<ul>\n";

    for (my $i=0; $i < @topicOrder; $i++) {
    
	next if ($topicOrder[$i] =~ m/selected/ ||
		 $topicOrder[$i] =~ m/ignore/);

	print "<li><a href=\"\#$topicOrder[$i]\">$topics{$topicOrder[$i]}</a></li>\n";

    }

    print "</ul>\n</div>\n";

    for (my $i=0; $i < @topicOrder; $i++) {
    
	next if ($topicOrder[$i] =~ m/selected/ ||
		 $topicOrder[$i] =~ m/ignore/);

	print "<h2 id=\"$topicOrder[$i]\">$topics{$topicOrder[$i]}</h2>\n";
	print '<table>
';

	for (my $paperId=0; $paperId<@allEntries; $paperId++) {

	    if ($allEntries[$paperId]{topics} =~ m/$topicOrder[$i]/) {

		#lets ignore papers marked as ignore
		next if ($allEntries[$paperId]{topics} =~ m/ignore/);
	
		print "<tr><td></td><td class=\"paper-entry\">";
		PrintPaper($paperId);
		print "</td></tr>";
	    }	
	}

	print "</table>\n";
    }
}

sub PrintByTime() {

    my $lastPrintedYear = -1;
    
    for (my $i=0; $i<@allEntries; $i++) {
	my $year = $allEntries[$i]{year};
	if ($year != $lastPrintedYear) {

	    #end the previous table
	    print "</table>\n" if ($lastPrintedYear != -1);

	    print "<div class=\"year-header\">$year</div>\n";
	    print '<table>
';
	    
	    $lastPrintedYear = $year;
	}

	#lets ignore papers marked as ignore
	##print STDERR $allEntries[$i]{URL};
	next if ($allEntries[$i]{topics} =~ m/ignore/);
	
	print "<tr><td></td><td class=\"paper-entry\">";
	PrintPaper($i);
	print "</td></tr>";
    }
    
}


sub PrintSelected() {
    
    die "print selected not yet implemented\n";
}

sub PrintCv()  {

for (my $i=0; $i < @topicOrder; $i++) {
    
	next if ($topicOrder[$i] =~ m/selected/ ||
		 $topicOrder[$i] =~ m/ignore/);

	print "\\item \{\\bf $topics{$topicOrder[$i]}\}\n";
	print "\\begin\{innerlist\}\n";

	for (my $paperId=0; $paperId<@allEntries; $paperId++) {

		if ($allEntries[$paperId]{topics} =~ m/$topicOrder[$i]/) {

			#lets ignore papers marked as ignore or 
			next if ($allEntries[$paperId]{topics} =~ m/ignore/);
			#lets ignore papers of type misc
			next if ($allEntries[$paperId]{paperType} =~ m/misc/);

			my $author = $allEntries[$paperId]{author};
	        $author =~ s/ and /, /g;

	        my $venue = "unknown";
	        if (defined($allEntries[$paperId]{booktitle})) {
		        $venue = $allEntries[$paperId]{booktitle};
	        }
	        elsif (defined($allEntries[$paperId]{journal})) {
		        $venue = $allEntries[$paperId]{journal};
	        }
	        elsif (defined($allEntries[$paperId]{howpublished})) {
		        $venue = $allEntries[$paperId]{howpublished};
	        }
	        else {
		       die "Could not find venue for $allEntries[$paperId]{title}\n";
	        }

	        my $title = $allEntries[$paperId]{title};
	        $title =~ s/(\{|\})//g;

	        print "\\item\n";

	        if (defined $allEntries[$paperId]{impact}) {
	        	print "\\hspace{-0.15in} * ";
	        }

	        print "    $author,\n";
	        print "    \`\`$title,\"\n";
	        print "     \{\\em $venue,\}\n";

	        if ($allEntries[$paperId]{paperType} =~ m/article/) {
	        	print "    $allEntries[$paperId]{volume}($allEntries[$paperId]{number},\n";
	        }

	        print "    $allEntries[$paperId]{month} $allEntries[$paperId]{year}.\n";

	        if (defined $allEntries[$paperId]{acceptancerate}) {
		      print "    ($allEntries[$paperId]{acceptancerate}\\\% acceptance rate)\n";
	        }	

	        if (defined $allEntries[$paperId]{note}) {
		      print "    \{\\bf $allEntries[$paperId]{note}\}\n";
	        }	

		}
    }
    print "\\end\{innerlist\}\n";
}
}

sub PrintPaper() {
    my ($paperId) = @_;

    my $author = $allEntries[$paperId]{author};
    $author =~ s/ and /, /g;

    my $venue = "unknown";
    if (defined($allEntries[$paperId]{booktitle})) {
	$venue = $allEntries[$paperId]{booktitle};
    }
    elsif (defined($allEntries[$paperId]{journal})) {
	$venue = $allEntries[$paperId]{journal};
    }
    elsif (defined($allEntries[$paperId]{howpublished})) {
	$venue = $allEntries[$paperId]{howpublished};
    }
    else {
	die "Could not find venue for $allEntries[$paperId]{title}\n";
    }

    my $title = $allEntries[$paperId]{title};
    $title =~ s/(\{|\})//g;

    print "<div class=\"paper-title\"><a href=\"$allEntries[$paperId]{URL}\">$title</a></div>\n";
    print "<div class=\"paper-authors\">$author</div>\n";
    print "<div class=\"paper-venue\">$venue, $allEntries[$paperId]{year}</div>\n";

    if (defined $allEntries[$paperId]{note}) {
	print "<div class=\"paper-note\">$allEntries[$paperId]{note}</div>\n";
    }

    if (defined $allEntries[$paperId]{resource}) {
	print "<div class=\"paper-resource\">$allEntries[$paperId]{resource}</div>\n";
    }
}


################ main #############

foreach my $file (@files) {
    
    open (my $F, $file) or die;

    while (<$F>) {

	#ignore comments and empty lines
	if (m/^\@comment/ || 
	    m/^\s*$/) {
	
	}
	#topic
	elsif (m/^([a-z]+)\s+:: (.+)$/) {
	    my ($topicKey, $topic) = ($1, $2);
	    $topics{$topicKey} = $topic;
	    push @topicOrder, $topicKey;

	}
    #start of a new entry
    elsif (m/^@([a-z]+)\{(.+)\,/) {
    	my ($paperType, $paperKey) = ($1, $2);    	
        $count++;
        $allEntries[$count]{paperType}=$paperType;
        $allEntries[$count]{paperKey}=$paperKey;
    }
	elsif (m/\s+([a-zA-Z]+)=\{(.+)\}/) {
	    my ($key, $value) = ($1, $2);

	    if (defined($allEntries[$count]) && 
		    defined($allEntries[$count]{$key})) {		
   		   die "duplicate $count or $count/$key";
	    }
	    
	    $allEntries[$count]{$key} = $value;
	}
    }
}


if ($MODE eq "time") {
    PrintHeader();
    PrintByTime();
    PrintFooter();
}
elsif ($MODE eq "topic") {
    PrintHeader();
    PrintByTopic();
    PrintFooter();
}
elsif ($MODE eq "selected") {
    PrintHeader();
    PrintSelected();
    PrintFooter();
}
elsif ($MODE eq "cv") {
	PrintCv();
}
else {
    die "unknown mode: $MODE\n";
}

File: pubs.py
import sys
import re
from collections import defaultdict
import configargparse
from typing import Dict, List

UW_STUDENTS = [
    "Xieyang Xu",
    "Tapan Chugh",
    "Xiangfeng Zhu",
    "Weixin Deng",
    "Anish Nyayachavadi",
    "Shubham Tiwari",
    "Wei Shen",
    "Brody Frank",
    "Natalie Neamtu",
    "Yuyao Wang",
]

POSTDOCS = [
    "Xiangyu Gao",
    "Sam Kumar",
    "Klaus-Tycho Forster",
    "Monia Ghobadi",
    "Hongqiang Liu",
]

MSR_STUDENTS = [
    "Nick Giannarakis",
    "Danyang Zhuo",
    "Arpit Gupta",
    "Aaron Gember-Jackson",
    "Ryan Beckett",
    "Seyed Fayaz",
    "Raajay Vishwanathan",
    "Xin Jin",
    "Hongqiang Liu",
    "Peng Sun",
    "Ari Fogel",
    "Rayman Preet Singh",
    "Luis Pedrosa",
    "Vijay Adhikari",
    "Yingying Chen",
    "Chun-Te Chu",
    "Chi-Yao Hong",
    "Jason Croft",
    "Colin Dixon",
    "Hossein Falaki",
    "Radhika Niranjan Mysore",
    "Patrick Verkaik",
    "Aruna Balasubramanian",
    "Lindsey Poole",
    "Krishna Ramachandran",
]


def print_header():
    print(
        """
<html>
<head>
  <title>ratul's publications</title>
</head>
<body bgcolor="#ffffff">
"""
    )


def print_footer():
    print(
        """
</body>
</html>
"""
    )


def print_by_topic(topic_order, topics, all_entries):
    print("<ul>")
    for topic in topic_order:
        if re.search(r"selected|ignore", topic):
            continue
        print(f'<li><a href="#{topic}">{topics[topic]}</a>')
    print("</ul>")

    for topic in topic_order:
        if re.search(r"selected|ignore", topic):
            continue
        print(f"<p><a name={topic}></a>\n<b>{topics[topic]}</b>")
        print(
            """
<table style="width: 100%;" cellpadding="2" cellspacing="2">
<tr><td width=10px><td></tr>
"""
        )
        for entry in all_entries:
            if re.search(topic, entry["topics"]):
                if re.search(r"ignore", entry["topics"]):
                    continue
                print_paper(entry)
        print("</table>")


def print_by_time(all_entries):
    last_printed_year = -1
    for entry in all_entries:
        year = entry["year"]
        if year != last_printed_year:
            print("</table>")
            print(f"<p><b>{year}</b>")
            print(
                """
<table style="width: 100%;" cellpadding="2" cellspacing="2">
<tr><td width=10px><td></tr>
"""
            )
            last_printed_year = year
        if re.search(r"ignore", entry["topics"]):
            continue
        print_paper(entry)


def print_pubtype(all_entries: List[Dict[str, str]]) -> None:
    pub_types = defaultdict(lambda: [])

    for entry in all_entries:
        pub_types[entry["pubtype"]].append(entry)

    print(f"Found pubtypes: {pub_types.keys()}", file=sys.stderr)

    for pub_type, entries in pub_types.items():
        print(f"<p><b>{pub_type}</b><br>")
        for entry in entries:
            if re.search(r"ignore", entry["topics"]):
                continue
            print_paper_pubtype(pub_type, entry)


def print_paper_pubtype(pub_type: str, entry: Dict[str, str]) -> None:
    print(f"printing {entry['paperKey']}", file=sys.stderr)

    print("<p>")
    if "impact" in entry:
        print("<sup>*</sup>", end="")

    authors = [author for author in entry["author"].split(" and ")]
    for index, author in enumerate(authors):
        if index == len(authors) - 1 and len(authors) != 1:
            # last author?
            print(f" and ", end="")
        else:
            print(f" ", end="")

        print(f"{author}", end="")

        # add a comma if we have more than two authors and this is not the last author
        if index < len(author) - 1 and len(author) > 2:
            print(",", end="")

        if author in UW_STUDENTS or author in MSR_STUDENTS:
            print("<sup>1</sup>", end="")

        if author in POSTDOCS:
            print("<sup>2</sup>", end="")

    title = re.sub(r"(\{|\})", "", entry["title"])
    print(f"    ``{title}''", end="")

    venue = (
        entry.get("booktitle")
        or entry.get("journal")
        or entry.get("howpublished")
        or "unknown"
    )
    print(f", <i>{venue}</i>", end="")

    if re.search(r"article", entry["paperType"]):
        print(f", {entry['volume']}({entry['number']})", end="")
    if "pages" in entry:
        print(f", pages {entry['pages']}", end="")
    print(f", {entry['month']} {entry['year']}.", end="")

    if "acceptancerate" in entry:
        print(f"    ({entry['acceptancerate']}% acceptance rate)", end="")
    if "citations" in entry:
        print(f"    ({entry['citations']} citations)", end="")
    if "note" in entry:
        print(f"    <b>({entry['note']})</b>")


def print_cv(topic_order, topics, all_entries):
    for topic in topic_order:
        if re.search(r"selected|ignore", topic):
            continue
        print(f"\\item {{\\bf {topics[topic]}}}")
        print("\\begin{innerlist}")
        for entry in all_entries:
            if re.search(topic, entry["topics"]):
                if re.search(r"ignore", entry["topics"]) or re.search(
                    r"misc", entry["paperType"]
                ):
                    continue
                author = entry["author"].replace(" and ", ", ")
                venue = (
                    entry.get("booktitle")
                    or entry.get("journal")
                    or entry.get("howpublished")
                    or "unknown"
                )
                title = re.sub(r"(\{|\})", "", entry["title"])
                print("\\item")
                if "impact" in entry:
                    print("\\hspace{-0.15in} * ")
                print(f"    {author},")
                print(f"    ``{title},''")
                print(f"     {{\\em {venue},}}")
                if re.search(r"article", entry["paperType"]):
                    print(f"    {entry['volume']}({entry['number']},")
                print(f"    {entry['month']} {entry['year']}.")
                if "acceptancerate" in entry:
                    print(f"    ({entry['acceptancerate']}\\% acceptance rate)")
                if "note" in entry:
                    print(f"    {{\\bf {entry['note']}}}")
        print("\\end{innerlist}")


def print_paper(entry):
    author = entry["author"].replace(" and ", ", ")
    venue = (
        entry.get("booktitle")
        or entry.get("journal")
        or entry.get("howpublished")
        or "unknown"
    )
    title = re.sub(r"(\{|\})", "", entry["title"])
    print(
        f"""
<a href={entry['URL']}><b>{title}</b></a><br>
{author}<br>
{venue}, {entry['year']}<br>
"""
    )
    if "note" in entry:
        print(f"<b>{entry['note']}</b><br>")
    if "resource" in entry:
        print(f"{entry['resource']}<br>")
    print("<br>")


def main():
    parser = configargparse.ArgParser(description="Process publication data.")
    parser.add_argument("--verbose", action="store_true", help="debug mode [off]")
    parser.add_argument(
        "--mode", type=str, default="time", choices=["time", "topic", "cv", "pubtype"]
    )
    parser.add_argument("files", nargs="+", help="list of files")
    args = parser.parse_args()

    topic_order = []
    topics = {}
    all_entries = []

    for file in args.files:
        with open(file, "r") as f:
            for line in f:
                if re.match(r"^\@comment", line) or re.match(r"^\s*$", line):
                    # ignore comments and blank lines
                    continue
                elif match := re.match(r"^([a-z]+)\s+:: (.+)$", line):
                    # topic lines at the top of the file
                    topic_key, topic = match.groups()
                    topics[topic_key] = topic
                    topic_order.append(topic_key)
                elif match := re.match(r"^\@([a-z]+)\{(.+)\,", line):
                    # the first line of an entry
                    paper_type, paper_key = match.groups()
                    all_entries.append({"paperType": paper_type, "paperKey": paper_key})
                elif match := re.match(r"\s+([a-zA-Z]+)=\{(.+)\}", line):
                    # following up lines of an entry
                    key, value = match.groups()
                    if key in all_entries[-1]:
                        raise ValueError(
                            f"duplicate {key=} in {all_entries[-1]['paperKey']}"
                        )
                    all_entries[-1][key] = value

    if args.mode == "time":
        print_header()
        print_by_time(all_entries)
        print_footer()
    elif args.mode == "topic":
        print_header()
        print_by_topic(topic_order, topics, all_entries)
        print_footer()
    elif args.mode == "cv":
        print_cv(topic_order, topics, all_entries)
    elif args.mode == "pubtype":
        print_header()
        print_pubtype(all_entries)
        print_footer()
    else:
        raise ValueError(f"unknown mode: {args.mode}")


if __name__ == "__main__":
    main()
